highfidelity/hifi
# Fade on model items and avatars

This PR adds fade events using a new transition framework that implements in a more usable manner the prototype dissolve fade effect as presented in WL branch 21378. The currently implemented and working events are:
- Model based entity items when they appear for the first time in a location.
- Avatars when they enter or leave a location.

# Under the hood
## Gradients
The fade effect is based on the thresholding of a gradient in 3D space. The gradient has values between 0 and 1. When the gradient has a value beneath a user defined threshold, the object is transparent. This gradient is the sum of a base radial gradient centered on the object's position (with value 0 at this position) and a noise gradient. Each base and noise gradient level can be modified as well as the size of these gradients in the three dimensions. The center of the base gradient is furthermore offset in position depending on the type of fade event.

The threshold result can be inverted with the "Invert" check box in which case the transparent pixels correspond to gradient values above the threshold.

## Edges
The outer edge of the fade effect is the limit between transparent and opaque pixels. The inner edge is set with the "edge width" parameter, between 0 and 1 as a ratio of the gradient size. Both edges have separate colors and color intensities, in the emissive channel.

# Fade editing
The debugFade.js script in scripts\developer\utilities\render can be used to tweak the fade effect on the 4 supported events, the above two being the only ones processed in the application at this time. Checking the "Edit Fade" checkbox applies cyclicly the selected fade event to the object in front of you. To manually control the threshold value, click on "Manual" under "Edit Fade". In this mode, the "Threshold" slider controls the value.

All changes to the various fade parameters apply to the currently selected event.

## Save
Clicking on the save button saves a JSON configuration file under interface/resources/config with name corresponding to the currently edited event. This file contains the current settings only for the edited event.

## Load
Clicking on the load button loads back the settings stored in the JSON configuration file under interface/resources/config with name corresponding to the currently edited event.

**Note**: the JSON configuration files are not automatically loaded at launch time.

# TEST PLAN
1. Check that each model based entity that appears in the world has a correct dissolve fade event applied to it.
2. When switching back and forth between two locations, the model based entity should have a fade event triggered each time.
3. Check that each avatar that appears in a location has a correct fade event applied to it, different from the model based entity.
4. Check that each avatar (except your avatar) that leaves a location has a correct fade event applied to it that is the inverse of the "avatar appear" event.
5. With the "debugFade.js" script running, checking the "Edit Fade" checkbox should apply the event selected to the combobox at the right side of the checkbox on the object in front of you.
6. With the "debugFade.js" script running, and the "Edit Fade" checkbox checked, modifying any parameter of the edit window should modify the fade event accordingly.
7. With the "debugFade.js" script running, and the "Edit Fade" checkbox checked, switching to another selected event should modify the fade's visual appearance.
7. With the "debugFade.js" script running, and the "Edit Fade" checkbox checked, switching to another object by changing the point of view should remove the fade event from the previous object and apply it to the new one.
8. With the "debugFade.js" script running, and both the "Edit Fade" and "Manual" checkboxes checked, moving the "Threshold" slider should modify the fade gradient's threshold value. 
9. With the "debugFade.js" script running, if the "Edit Fade" checkbox was checked, unchecking it should remove the fade event from the object in front of you
10. When an item has a fade event applied to it, it should be rendered "double-sided"
